infrastructure/src/main/java/org/corfudb/infrastructure/health/HealthReport.java

```diff
@@ -0,0 +1,193 @@
+                createRuntimeReportedHealthStatus(componentHealthStatusSnapshot);
+        String overallReason;
+        ComponentStatus overallStatus;
+        if (initReportedHealthStatus.isEmpty()) {
+            overallStatus = UNKNOWN;
+            overallReason = OVERALL_STATUS_UNKNOWN;
+        } else if (!isHealthy(initReportedHealthStatus)) {
+            overallStatus = DOWN;
+            overallReason = OVERALL_STATUS_DOWN;
+        } else if (!isHealthy(runtimeReportedHealthStatus)) {
+            overallStatus = FAILURE;
+            overallReason = OVERALL_STATUS_FAILURE;
+        } else {
+            overallStatus = UP;
+            overallReason = OVERALL_STATUS_UP;
+        }
+        return HealthReport.builder()
+                .status(overallStatus)
+                .reason(overallReason)
+                .init(initReportedHealthStatus)
+                .runtime(runtimeReportedHealthStatus)
+                .build();
+    }
+
+    /**
+     * @return A json representation of a health report
+     */
+    public String asJson() {
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        return gson.toJson(this);
+    }
+
+    private static boolean isHealthy(Set<ComponentReportedHealthStatus> componentHealthMap) {
+        return !componentHealthMap.isEmpty() && componentHealthMap.stream()
+                .allMatch(healthStatus -> healthStatus.status == UP);
+    }
+
+    private static Set<ComponentReportedHealthStatus> createInitReportedHealthStatus(Map<Component,
+            HealthStatus> componentHealthStatus) {
+        return componentHealthStatus.entrySet().stream().map(entry -> {
+            final Component component = entry.getKey();
+            final HealthStatus healthStatus = entry.getValue();
+            if (healthStatus.isInitHealthy()) {
+                return new ComponentReportedHealthStatus(component, UP,
+                        COMPONENT_INITIALIZED);
+            } else {
+                return new ComponentReportedHealthStatus(component, DOWN,
+                        COMPONENT_NOT_INITIALIZED);
+            }
+        }).collect(ImmutableSet.toImmutableSet());
+    }
+
+    private static Set<ComponentReportedHealthStatus> createRuntimeReportedHealthStatus(Map<Component,
+            HealthStatus> componentHealthStatus) {
+        return componentHealthStatus.entrySet().stream().map(entry -> {
+            final Component component = entry.getKey();
+            final HealthStatus healthStatus = entry.getValue();
+            if (healthStatus.getLatestRuntimeIssue().isPresent()) {
+                Issue issue = healthStatus.getLatestRuntimeIssue().get();
+                return new ComponentReportedHealthStatus(component, FAILURE, issue.getDescription());
+            } else if (!healthStatus.isRuntimeHealthy()) {
+                return new ComponentReportedHealthStatus(component, DOWN, COMPONENT_IS_NOT_RUNNING);
```

Comment:
It was required that we have these statuses. 
UNKNOWN means we don't know what components are there because none registered to be initialized. 
DOWN means the component is not working because it was not initialized or got shut down.
UP means it's running correctly and has no failures. 
FAILURE means it's UP but encountered at least one runtime issue, which is not yet resolved.

Reply:
I see. If this is a requirement from NSX, probably we can add a comment to explain the `FAILURE` in corfu.